openpype/hosts/cinema4d/plugins/publish/collect_instances.py

```diff
@@ -0,0 +1,185 @@
+
+class CollectInstances(pyblish.api.ContextPlugin):
+    """Gather instances by objectSet and pre-defined attribute
+
+    This collector takes into account assets that are associated with
+    an objectSet and marked with a unique identifier;
+
+    Identifier:
+        id (str): "pyblish.avalon.instance"
+
+    Limitations:
+        - Does not take into account nodes connected to those
+            within an objectSet. Extractors are assumed to export
+            with history preserved, but this limits what they will
+            be able to achieve and the amount of data available
+            to validators. An additional collector could also
+            append this input data into the instance, as we do
+            for `pype.rig` with collect_history.
+
+    """
+
+    label = "Collect Instances"
+    order = pyblish.api.CollectorOrder
+    hosts = ["cinema4d"]
+
+    def process(self, context, doc=None):
+        if not doc:
+            doc = c4d.documents.GetActiveDocument()
+
+        objectset = set()
+        for obj in lib.walk_hierarchy(doc.GetFirstObject()):
+            if obj.attrs.get("id"):
+                objectset.add(obj.attrs)
+
+
+        ctx_frame_start = context.data['frameStart']
+        ctx_frame_end = context.data['frameEnd']
+        ctx_handle_start = context.data['handleStart']
+        ctx_handle_end = context.data['handleEnd']
+        ctx_frame_start_handle = context.data['frameStartHandle']
+        ctx_frame_end_handle = context.data['frameEndHandle']
+
+        context.data['objectsets'] = objectset
+        for objset in objectset:
+
+            if not objset.get("id"):
+                continue
+
+            if objset.get("id") != "pyblish.avalon.instance":
+                continue
+
+            # The developer is responsible for specifying
+            # the family of each instance.
+            has_family = objset.get("family")
+            assert has_family, "\"%s\" was missing a family" % objset
+
+            members = objset.get("SELECTIONOBJECT_LIST")
+            if members is None:
+                self.log.warning("Skipped empty instance: \"%s\" " % objset)
+                continue
+
+            self.log.info("Creating instance for {}".format(lib.serialize_c4d_data(objset.op)))
+
+            data = dict()
+
+            for key, value in objset:
+                data[key] = lib.serialize_c4d_data(value)
+
+            # temporarily translation of `active` to `publish` till issue has
+            # been resolved, https://github.com/pyblish/pyblish-base/issues/307
+            if "active" in data:
+                data["publish"] = data["active"]
+
+            # Collect members - c4d.BaseObject
+            members = [lib.ObjectPath(obj=members.ObjectFromIndex(doc, idx)) for idx in range(members.GetObjectCount())]
+
+
+            # Collect Children - c4d.BaseObject
+            children = [lib.ObjectPath(obj=obj) for member in members for obj in lib.walk_hierarchy(member.obj.GetDown())]
```

Comment:
Note - I know NOTHING of Cinema4D but I had a quick glance over [this page](https://developers.maxon.net/docs/Cinema4DPythonSDK/html/modules/c4d/C4DAtom/GeListNode/index.html?highlight=getdown#GeListNode.GetChildren).
Wouldn't `GetDown` just return the "first child" of each node? Shouldn't this be `GetChildren()`?

It seems `walk_hierarchy` does seem to work around that with `GetNext()` - but wouldn't starting from the child only walk the hierarchy of those children. I feel the logic is a bit off somewhere.

Is it `walk_hierarchy` that is also walking the siblings of the object passed into it? I wouldn't expect that myself. I would've expected it to walk down the hierarchy from the node passed in and not touch its direct siblings.

Reply:
Should `children` include all descendants in the hierarchy or just the direct children of the object? I was collecting all children, `GetChildren()` is only direct children.

I see what your saying with `walk_hierarchy` not working as you would expect. There is a reason and I don't know if it is a good one. In C4D the pattern for listing most things in a project file is to get the first object, like `doc.GetFirstMaterial()` or `doc.GetFirstObject()`, which then requires something like:

 ```
obj = doc.GetFirstObject()
while obj:
    #do stuff with obj
    obj = obj.GetNext()
```

What I should probably do is just make a `get_siblings()` function so that `walk_hierarchy` operates more like `os.walk`

Reply:
> Should children include all descendants in the hierarchy or just the direct children of the object? I was collecting all children, GetChildren() is only direct children.

Your assumption is correct - it would be ALL children + grandchildren.

> I see what your saying with walk_hierarchy not working as you would expect. There is a reason and I don't know if it is a good one. In C4D the pattern for listing most things in a project file is to get the first object, like doc.GetFirstMaterial() or doc.GetFirstObject(), which then requires something like:

Sort of, yes. I read the code as it currently doing this:

```python
current_node = member.obj
first_child_of_current_node = member.obj.GetDown()
all_siblings_and_its_children_and_grandchildren = walk_hierarchy(first_child_of_current_node)
```

Which if that is the case then functionality wise it does do as it should,

However solely based on function name I wouldn't have expected `walk_hierarchy` to also go over "siblings" of the input object but would have instead expected it to just go down, and go from there. But I can see from the C4D logic point of view how `walk_hierarchy` with just the `.Next()` implementation is much more generic maybe. I somehow would've expected `walk_hierarchy` to first do a `.GetDown()` internally so that it's really only about walking down the hierarchy and its children's siblings, etc.